services/venice-server/src/main/java/com/linkedin/venice/grpc/GrpcRequestContext.java

```diff
@@ -0,0 +1,93 @@
+package com.linkedin.venice.grpc;
+
+import com.linkedin.davinci.listener.response.ReadResponse;
+import com.linkedin.venice.listener.ServerStatsContext;
+import com.linkedin.venice.listener.request.RouterRequest;
+import com.linkedin.venice.response.VeniceReadResponseStatus;
+import io.grpc.stub.StreamObserver;
+
+
+/**
+ * We need to keep track of each request as it goes through the pipeline so that we can record the necessary metrics
+ * and separate different parts of the logic for the response. If a specific handler raises an error, we set
+ * the hasError flag to true and stop executing the rest of the pipeline excluding the stats collection.
+ */
+public class GrpcRequestContext<T> {
+  private final StreamObserver<T> responseObserver;
+  private final Class<?> responseType;
+  private final ServerStatsContext serverStatsContext;
+
+  private RouterRequest routerRequest;
+  private ReadResponse readResponse = null;
+  private VeniceReadResponseStatus readResponseStatus;
+  private String errorMessage;
```

Comment:
Currently `ReadResponse` is only used for IO requests. For all requests and for early errored requests in Netty we use `HttpShortcutResponse`.  

Ideally it would be good to have common request context for all requests but I think I'll skip it for now